### ISSUE TITLE
Misleading error for #! not being on the first line and column

### ISSUE THREAD

**Issue by jjonescz** [dotnet/roslyn#83111]
**Version Used**: 5.7.0

**Steps to Reproduce**:

```cs
class Foo { }
#!/usr/bin/env dotnet
class Program { static void Main() { } }
```

with `<Features>FileBasedProgram</Features>`

**Expected Behavior**: Error about `#!` not being on the first line.

**Actual Behavior**: `(2,1): error CS1040: Preprocessor directives must appear as the first non-whitespace character on a line`

This error doesn't make sense - the directive is the first character on its line.

Originally reported by @333fred in https://github.com/dotnet/sdk/pull/53614#discussion_r3054641418.
